UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,627 @@
+              PlatformAcpiTable->Rsdp            = (EFI_PHYSICAL_ADDRESS)(UINTN)StartAddress;
+              PlatformAcpiTable->Header.Revision = UNIVERSAL_PAYLOAD_ACPI_TABLE_REVISION;
+              PlatformAcpiTable->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_ACPI_TABLE);
+            }
+          } else if (AsciiStrnCmp (TempStr, "acpi-nvs", AsciiStrLen ("acpi-nvs")) == 0) {
+            DEBUG ((DEBUG_INFO, "  acpi-nvs"));
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiACPIMemoryNVS);
+          } else if (AsciiStrnCmp (TempStr, "smbios", AsciiStrLen ("smbios")) == 0) {
+            DEBUG ((DEBUG_INFO, " build smbios, NumberOfBytes:%x", NumberOfBytes));
+            SmbiosTable = BuildGuidHob (&gUniversalPayloadSmbios3TableGuid, sizeof (UNIVERSAL_PAYLOAD_SMBIOS_TABLE) + sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT));
+            if (SmbiosTable != NULL) {
+              SmbiosTable->Header.Revision  = UNIVERSAL_PAYLOAD_SMBIOS_TABLE_REVISION;
+              SmbiosTable->Header.Length    = sizeof (UNIVERSAL_PAYLOAD_SMBIOS_TABLE);
+              SmbiosTable->SmBiosEntryPoint = (EFI_PHYSICAL_ADDRESS)(UINTN)(&StartAddress);
+            }
+          }
+        }
+      }
+    } // end of reserved-memory
+
+    if (AsciiStrnCmp (NodePtr->Name, "serial@", AsciiStrLen ("serial@")) == 0) {
+      //
+      // Create SerialPortInfo HOB.
+      //
+      Serial = BuildGuidHob (&gUniversalPayloadSerialPortInfoGuid, sizeof (UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO));
+      ASSERT (Serial != NULL);
+      if (Serial == NULL) {
+        break;
+      }
+
+      Serial->Header.Revision = UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO_REVISION;
+      Serial->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO);
+      Serial->RegisterStride  = 1;
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+        if (AsciiStrCmp (TempStr, "current-speed") == 0) {
+          Data32 = (UINT32 *)(PropertyPtr->Data);
+          DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+
+          Serial->BaudRate = Fdt32ToCpu (*Data32);
+        } else if (AsciiStrCmp (TempStr, "reg") == 0) {
+          Data32       = (UINT32 *)(PropertyPtr->Data);
+          Attribute    = Fdt32ToCpu (*(Data32 + 0));
+          StartAddress = Fdt32ToCpu (*(Data32 + 2));
+          DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          DEBUG ((DEBUG_INFO, " StartAddress   %016lX\n", StartAddress));
+          DEBUG ((DEBUG_INFO, " Attribute      %016lX\n", Attribute));
+
+          Serial->RegisterBase = StartAddress;
+          Serial->UseMmio      = (Attribute&SS_32BIT_MEMORY_SPACE) == SS_32BIT_MEMORY_SPACE ? TRUE : FALSE;

```

Comment:
per today's UPL community meeting, we would like to align serial node with existing isa-bus schema (extend to lpc and espi bus too).
isa-bus schema: https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/isa/isa-bus.yaml

So there will be 2 types of serial device:
1. if serial device under isa/lpc/espi: first element of reg property will be 1 to indicate it is IO, otherwise it will be MMIO.
2. if serial device under pci: follow address-cells=<3> case that reg attribute field will indicate the type.

example:
espi@1f,0 {
				#address-cells = <2>;
				#size-cells = <1>;
				compatible = "isa";
				reg = <0xf800 0x0 0x0 0x0 0x0>;
				ranges = <1 0 0 0 0 0x100>;

				serial@3F8 {
                                  compatible = "isa";
                                  reg-io-width = <0x00000001>;
                                  reg = <1 0x3f8 8>;
                                  current-speed = <0x0001c200>;
				};

Reply:
we will implement one platform example for the first version.

Reply:
resolved.